kernel/kernel-api/src/main/java/io/delta/kernel/internal/stats/FileSizeHistogram.java

```diff
@@ -0,0 +1,222 @@
+/*
+ * Copyright (2024) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.delta.kernel.internal.stats;
+
+import static io.delta.kernel.internal.util.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+import java.util.*;
+
+/** A histogram that tracks file size distributions and their counts. */
+public class FileSizeHistogram {
```

Comment:
More thoughts:
1. can't `fileCounts` and `totalBytes` be `long` arrays? Do we ever need to hold null?
2. I think `sortedBinBoundaries` can be an array instead of a List? we never want to resize it, right? this is more memory efficient and lets us avoid boxing/unboxing the `Long`. You can use `Arrays.binarySearch` below etc

Reply:
Thanks for the suggestion! updated